File: backend/dataset/maxPathSum/maxPathSum_9.py
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        ## RC ##
        ## APPROACH : RECURSION ##
        ## LOGIC : POST ORDER CALCULATION ##
        
		## TIME COMPLEXITY : O(N) ##
		## SPACE COMPLEXITY : O(N) ##
        
        self.max_sum = float('-inf')
        self.calculateMaxPath(root)
        return self.max_sum
        
    def calculateMaxPath(self,node):
        if(node):
            left = right = 0
            if(node.left):
                left = max(self.calculateMaxPath(node.left), 0)
            if(node.right):
                right = max(self.calculateMaxPath(node.right), 0)
            self.max_sum = max(left + right + node.val, self.max_sum)
            return max(left, right) + node.val

File: backend/CDataset/maxPathSum/maxPathSum_9.cpp
class Solution {    
public:
    class TreeNode{
        public:
        TreeNode* left;
        TreeNode* right;
        int val;
    };
    int maxPathSum( TreeNode* root ) {
        int maxSum = INT_MIN;
        helper( root, maxSum );
        return maxSum;
    }
    int helper( TreeNode* node, int &maxSum ) {
        if( !node ) return 0;
        int left = 0, right = 0;
        if( node->left )  left = helper( node->left, maxSum );
        if( node->right ) right = helper( node->right, maxSum );
        
        // node+left or node+right or node
        int leftOrRightOrNode = max( node->val, node->val + max( left, right ) );
        // set maxSum to max of maxSum, node+left+right, node+left, node+right, node
        maxSum = max( maxSum, max( node->val + left + right, leftOrRightOrNode ) );        
        return leftOrRightOrNode;
    }
};
